File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        // algorithm idea we store the max of sum of the nodes below
        // and that node itself. 
        // so basically we start at root and go left and go right. 
        // then keep going down till we hit leaf. once we hit leaf
        // store max sum as leaf value then go up. 
        // then chekc max as stated above.
        int max_sum = INT_MIN;
        maxSum(root, max_sum);
        return max_sum;
    }

    int maxSum(TreeNode* node, int& max_sum) {
        if (node == nullptr) return 0;
        int left = max(0, maxSum(node->left, max_sum));
        int right = max(0, maxSum(node->right, max_sum));

        max_sum = max(max_sum, node->val + left + right);
        return node->val + max(left, right);
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        max_length = 0
        hashset = set(s)
        for i in hashset:
            left, right = 0, 0
            count = 0
            while right < len(s):
                if s[right] != i:
                    count += 1
                if count <= k:
                    max_length = max(max_length, right - left + 1)
                    right += 1
                else:
                    while count > k:
                        if s[left] != i:
                            count -= 1
                        left += 1
                    right += 1
        return max_length